apps/web/src/views/GaugesVoting/hooks/useGaugesFilter.ts

```diff
@@ -51,6 +59,84 @@ export const useGaugesFilter = (fullGauges: Gauge[] | undefined) => {
     [filter],
   )
 
+  return {
+    searchText,
+    setSearchText,
+
+    filter,
+    setFilter,
+    onFilterChange,
+  }
+}
+
+const useGaugesFilterQueryState = () => {
+  const [searchText, setSearchText] = useQueryState(
+    'q',
+    parseAsString.withDefault('').withOptions({
+      history: 'replace',
+      shallow: true,
+    }),
+  )
+  const [filter, setFilter] = useQueryStates(
+    {
+      byChain: parseAsArrayOf(parseAsInteger).withDefault([]),
+      byFeeTier: parseAsArrayOf(
+        parseAsNumberLiteral<FeeAmount>([FeeAmount.LOW, FeeAmount.MEDIUM, FeeAmount.HIGH, FeeAmount.LOWEST]),
+      ).withDefault([]),
+      byType: parseAsArrayOf(parseAsStringLiteral<Gauges>([Gauges.Regular, Gauges.Boosted, Gauges.Capped])).withDefault(
+        [],
+      ),
+    },
+    {
+      history: 'replace',
+      shallow: true,
+    },
+  )
+
+  const onFilterChange = useCallback(
+    (type: OptionsType, value: FilterValue) => {
+      const opts = filter[type]
+      // select all
+      if (Array.isArray(value)) {
+        setFilter({
+          [type]: value.length === opts.length ? [] : value,
+        })
+      } else if (opts.some((x) => x === value)) {
+        // deselect one
+        setFilter({
+          [type]: opts.filter((v) => v !== value),
+        })
+      } else {
+        // select one
+        setFilter({
+          [type]: [...opts, value],
+        })
+      }
+    },
+    [filter, setFilter],
+  )
+
+  return {
+    searchText,
+    setSearchText,
+
+    filter,
+    setFilter,
+    onFilterChange,
```

Comment:
From the interface design perspective, it's better to return an "action" instead of an "event listener", because listeners are usually associated with user interactions, which better defined within the scope of component. To me the following interface would be more intuitive if combining all filters for gauges.
```typescript
const {
  searchText,
  setSearchText,
  chains,
  setChains,
  feeTiers,
  setFeeTiers,
  gaugeTypes,
  setGaugeTypes,
} = useGaugesFilters()
```

Reply:
may the name caused this confuse, 
renamed `onFilterChange` to `setFilter`, and hide previous `setFilter` as it never used outside

Reply:
From the single filter perspective, we can try to make sure that both state filter and query filter share the same interface. The benefit is that we only need to use the one we need and keep the peripheral logic the same

```typescript
// suppose we have two hooks `useQueryFilter` and `useStateFilter` who share the same interface
function useGaugesFilters(
  // We can pass in the hook as parameter to control the actual behavior
  useFilter,
) {
  const [searchText, setSearchText] = useFilter({ ... })
  ...
}

const { searchText: searchTextFromQuery, setSearchText: setSearchTextToQuery } = useGaugesFilters(useQueryFilter)

const { searchText, setSearchText } = useGaugesFilters(useStateFilter)
```